src/backend/executor/nodeHash.c

```diff
@@ -333,14 +341,40 @@ MultiExecParallelHash(HashState *node)
 										 &hashvalue, &hashkeys_null))
 					ExecParallelHashTableInsert(hashtable, slot, hashvalue);
 				hashtable->partialTuples++;
+
+				if (node->hs_quit_if_hashkeys_null && hashkeys_null)
+				{
+					/* CBDB_PARALLEL:
+					 * If we are LASJ and found NULL value by ourself or sibling processes had
+					 * found NULL values, quit and tell siblings to quit if possible.
+					 *
+					 * It's safe to fetch and set phs_lasj_has_null without lock here and at
+					 * other places. As it's a boolean and we don't need to have the most recent
+					 * value from CPU or Mem cache. And we should avoid more locks in HashJion Impl.
+					 * If we miss it here and some others set it at the same time, just bypass
+					 * and we may get it at the next Hash batch.
+					 * If we missed it across all batches, we will know it when PHJ_BUILD_HASHING_INNER
+					 * ends with the help of build_barrier.
+					 * If we never participated in building hash table, check it when hash table
+					 * creation job is finished.
+					 */
+					pstate->phs_lasj_has_null = true;
```

Comment:
Do we need to add barrier after updating phs_lasj_has_null?

Reply:
Done.